### PR TITLE
spo2:peripheral initialization and basic thread #22

### PR DESCRIPTION
**Pull request by reporiot**
Implementation of spo2 basic utility, see #22

### CODE REVIEW

File: include/ls_task.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * This struct serves as a shortcut for operating with FreeRTOS tasks such
+ * that it includes necessary fields for managing threads.
+ */
+struct _ls_task {
+	StaticTask_t data;
+	TaskHandle_t handle;
+	StackType_t stack[LS_TASK_STACK_SIZE];
+};
+
+typedef struct _ls_task ls_task;
```
Comment (nopeslide) [inline include/ls_task.h]: why not an anonymous struct directly typedefed?

File: include/ls_task.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#ifndef __LIFE_SENSOR_TASK__
+#define __LIFE_SENSOR_TASK__
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+
+#define LS_TASK_STACK_SIZE 0x1000
```
Comment (nopeslide) [inline include/ls_task.h]: i dislike the idea of a globally defined stacksize for assumably all tasks?

File: include/ls_task.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ */
+struct _ls_task {
+	StaticTask_t data;
+	TaskHandle_t handle;
+	StackType_t stack[LS_TASK_STACK_SIZE];
+};
+
+typedef struct _ls_task ls_task;
+
+#endif
```
Comment (nopeslide) [inline include/ls_task.h]: this file makes no sense in my opinion, for essentially saving few LoC

File: include/spo2.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * spo2() - SpO2 (Pulsoxy) meassurement device interface initialization
+ *
+ * spo2() initializes peripherals (4xADC, 1xPWM, 2xGPIO) that is necessary
+ * to operate together with the pulsoxy board which does preprocessing of input
+ * data. After initialization a thread is started which periodically reads in
+ * input data.
+ */
+void spo2(void);
```
Comment (nopeslide) [inline include/spo2.h]: bad function name. functions should be named after an action like `create_spo2_task`

File: src/spo2.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+
+#include "ls_task.h"
+#include "spo2.h"
+#include "spo2_driver.h"
+
+
+ls_task spo2_task;
```
Comment (nopeslide) [inline src/spo2.c]: would move this into `main.c`

File: src/spo2.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+static void spo2_read(uint32_t *data);
+
+/**
+ * print_values() - print aquired values
+ * @data:	space for sample readings
+ *
+ * Print values to serial out that have been aquired through the peripherals
+ * for demonstration and/or debug purpuses.
+ */
+static void print_values(uint32_t *data);
```
Comment (nopeslide) [inline src/spo2.c]: has nothing to do with spo2

File: include/spo2.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+#endif
```
Comment (nopeslide) [inline include/spo2.h:36]: `spo2` is a bad name, plz rename the file to sth like `task_spo2`

File: include/spo2.h [hunk 8]
```diff
@@ -0,0 +1,17 @@
+#ifndef __SPO2__
+#define __SPO2__
+
+#define SPO2_TASK_NAME "SpO2"
```
Comment (nopeslide) [inline include/spo2.h]: this define is not used in this header. so why is it here?

File: src/spo2.c [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	spo2_task.handle = xTaskCreateStatic(
+		do_spo2,
+		SPO2_TASK_NAME,
+		LS_TASK_STACK_SIZE,
+		NULL,
+		tskIDLE_PRIORITY,
+		spo2_task.stack,
+		&spo2_task.data
+	);
+}
```
Comment (nopeslide) [inline src/spo2.c]: the wrapper for task creation should be given the struct and the name,
even if its only created once, so it's following the style of multiple instanced processes and the name is defined in main.

File: src/spo2.c [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+
+/**
+ * do_spo2() - spo2 task worker function
+ * @pvParameters:	void* to pass function parameters.
+ *
+ * This worker function retrieves an input data sample from the ADCs. This
+ * is the place where further processing of the data will happen.
+ */
+static void do_spo2(void *pvParameters);
```
Comment (nopeslide) [inline src/spo2.c]: suddenly other naming scheme, how about `spo2_task`?

File: src/spo2.c [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ * for demonstration and/or debug purpuses.
+ */
+static void print_values(uint32_t *data);
+
+
+void
+spo2
+(void)
+{
+	spo2_init_peripherals();
```
Comment (nopeslide) [inline src/spo2.c:36]: why hide the init of each component in this wrapper?

File: src/spo2_driver.c [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+	.intr_type = GPIO_PIN_INTR_DISABLE
+};
+
+static gpio_config_t gpio_agc_inc_config = {
+	.pin_bit_mask = SPO2_AGC_INC_PIN,
+	.mode = GPIO_MODE_OUTPUT,
+	.pull_down_en = GPIO_PULLDOWN_ENABLE,
+	.pull_up_en = GPIO_PULLUP_DISABLE,
+	.intr_type = GPIO_PIN_INTR_DISABLE
+};
```
Comment (nopeslide) [inline src/spo2_driver.c:75]: are theses needed in ROM, can't they be put on stack?

File: src/spo2_driver.c [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+		data[0] = adc1_get_raw(SPO2_ADC_RED_DC);
+		data[1] = adc1_get_raw(SPO2_ADC_IRD_DC);
+		data[2] = adc1_get_raw(SPO2_ADC_RED_AC);
+		data[3] = adc1_get_raw(SPO2_ADC_IRD_AC);
+	}
+
+	data[0] /= SPO2_ADC_ROUNDS;
+	data[1] /= SPO2_ADC_ROUNDS;
+	data[2] /= SPO2_ADC_ROUNDS;
+	data[3] /= SPO2_ADC_ROUNDS;
```
Comment (nopeslide) [inline src/spo2_driver.c]: why no moving average?

**Review by nopeslide — CHANGES_REQUESTED** [review]

File: src/spo2_driver.c (hunk 12 above)
Comment (nopeslide) [reply src/spo2_driver.c:75]: unnecessary as it seems

Comment (reporiot) [reply src/spo2_driver.c:75]: stack won't be freed up. Resolving for now, until we have a better way to manage this.

File: src/spo2.c (hunk 6 above)
Comment (reporiot) [reply src/spo2.c]: Other than implementing another HAL, the only use of this read function is to request values that are necessary for calculating the spo2. Its purpose is not to get all values that are potentially available to the esp core.

Comment (reporiot) [reply src/spo2.c]: The print function is for demonstration purposes. We have to think of debug switches that compile various print functions into the binary on demand. I think there should not be any print function included into any stable binary (other than logging and similar).

Comment (ikstream) [reply src/spo2.c]: esp-idf provides a pretty useful library for outputting values. 
 https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/log.html

File: src/spo2_driver.c (hunk 13 above)
Comment (reporiot) [reply src/spo2_driver.c]: mavg will be implemented side by side when introducing ulp and ringbuffer

File: include/spo2.h (hunk 4 above)
Comment (reporiot) [reply include/spo2.h]: changed this to something better, but we need to discuss naming schemes

File: src/spo2.c (hunk 11 above)
Comment (reporiot) [reply src/spo2.c:36]: why should this pollute main? it's just encapsulation of driver code

**Review by ikstream — CHANGES_REQUESTED** [review]
Please have a look at [3.3.c](https://barrgroup.com/Embedded-Systems/Books/Embedded-C-Coding-Standard/White-Space-Rules/Blank-Lines)

**Review by ikstream — APPROVED** [review]
lgtm.

File: include/spo2.h [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * spo2_init() - SpO2 (Pulsoxy) meassurement device interface initialization
+ *
+ * spo2_init() initializes peripherals (4xADC, 1xPWM, 2xGPIO) that is necessary
+ * to operate together with the pulsoxy board which does preprocessing of input
+ * data. After initialization a thread is started which periodically reads in
+ * input data.
+ */
+void spo2_init(const char*, _spo2_task*);
```
Comment (nopeslide) [inline include/spo2.h]:
```suggestion
void spo2_init(const char *name, _spo2_task *spo2_task);
```
leaving out parameter names, which are also not documented, is a bad practice

File: include/spo2.h [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+ * @stack:	the address of the tasks' stack
+ *
+ * This struct serves as a shortcut for operating with FreeRTOS tasks such
+ * that it includes necessary fields for managing threads.
+ */
+typedef struct {
+	StaticTask_t tcb;
+	TaskHandle_t handle;
+	StackType_t stack[SPO2_STACK_SIZE];
+} _spo2_task;
```
Comment (nopeslide) [inline include/spo2.h:23]: no fan of names, which start with an underscore. Such names usually imply internal usage.

**Review by nopeslide — COMMENTED** [review]
May be merged, not happy with it, but as long no coding rules exist they can't be broken

**Review by nopeslide — APPROVED** [review]

**Pull request merged**